clippy_dev/src/sync.rs

```diff
@@ -0,0 +1,252 @@
+        .is_empty()
+    {
+        eprintln!("working directory must be clean before running `cargo dev sync pull`");
+        exit(1);
+    }
+}
+
+pub fn update_toolchain() {
+    let sh = Shell::new().expect("failed to create shell");
+    sh.change_dir(clippy_project_root());
+
+    assert_clean_repo(&sh);
+
+    // Update rust-toolchain file
+    let date = Utc::now().format("%Y-%m-%d").to_string();
+    update_lints::replace_region_in_file(
+        update_lints::UpdateMode::Change,
+        Path::new("rust-toolchain"),
+        "# begin autogenerated version\n",
+        "# end autogenerated version",
+        |res| {
+            writeln!(res, "channel = \"nightly-{date}\"").unwrap();
+        },
+    );
+
+    let message = format!("Bump nightly version -> {date}");
+    cmd!(sh, "git commit rust-toolchain --no-verify -m {message}")
+        .run()
+        .expect("FAILED to commit rust-toolchain file, something went wrong");
+}
+
+pub fn rustc_pull(hash: Option<String>) {
+    const MERGE_COMMIT_MESSAGE: &str = "Merge from rustc";
+
+    let sh = Shell::new().expect("failed to create shell");
+    sh.change_dir(clippy_project_root());
+
+    assert_clean_repo(&sh);
+
+    let date = Utc::now().format("%Y-%m-%d").to_string();
+    if hash.is_none()
+        && !std::fs::read_to_string("rust-toolchain")
+            .expect("failed to read rust-toolchain file")
+            .contains(&date)
+    {
+        eprintln!("Toolchain must be up to date before running `cargo dev sync pull`");
+        exit(1);
+    }
+
+    let commit = hash.unwrap_or_else(rustc_hash);
+
+    // Make sure josh is running in this scope
+    {
+        let _josh = start_josh();
+
+        // Fetch given rustc commit.
+        cmd!(
+            sh,
+            "git fetch http://localhost:{JOSH_PORT}/rust-lang/rust.git@{commit}{JOSH_FILTER}.git"
+        )
+        .run()
+        .expect(
+            "FAILED to fetch new commits, something went wrong (committing the rust-toolchain file has been undone)",
+        );
+    }
+
+    // This should not add any new root commits. So count those before and after merging.
+    let num_roots = || -> u32 {
+        cmd!(sh, "git rev-list HEAD --max-parents=0 --count")
+            .read()
+            .expect("failed to determine the number of root commits")
+            .parse::<u32>()
+            .unwrap()
+    };
+    let num_roots_before = num_roots();
+
+    // Merge the fetched commit.
+    cmd!(sh, "git merge FETCH_HEAD --no-verify --no-ff -m {MERGE_COMMIT_MESSAGE}")
+        .run()
+        .expect("FAILED to merge new commits, something went wrong");
+
+    // Check that the number of roots did not increase.
+    if num_roots() != num_roots_before {
+        eprintln!("Josh created a new root commit. This is probably not the history you want.");
+        exit(1);
+    }
+}
+
+pub fn rustc_push(rustc_path: String, github_user: &str) {
+    const BRANCH: &str = "clippy-subtree-update";
```

Comment:
At least for initial experimentation, it is quite useful to be able to overwrite the branch. May be worth to make this a flag with a default value.

Reply:
:shrug: up to you, for testing I sometimes want to have multiple branches to push to, to simulate "imagine we push to rustc, changes happen there, changes happen locally, we pull from rustc and then we push again".

Reply:
Makes sense. I'll add this then as a defaulted flag.